connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java

```diff
@@ -517,13 +517,13 @@ public void testTaskAssignmentWhenWorkerBounces() {
         leaderAssignment = deserializeAssignment(result, leaderId);
         assertAssignment(leaderId, offset,
                 Collections.emptyList(), 0,
-                Collections.emptyList(), 0,
+                Collections.emptyList(), 1,
```

Comment:
Isn't this a regression? We shouldn't be revoking tasks in this round since, without those revocations, we'd have a balanced assignment.

Reply:
Yeah that's true. Revoking tasks at this point leads to imbalance. This was happening since as per the new changes, the moment the delay expires we were allowing revocation. So, at this point, the flag is true, the delay goes to 0 and since revocation is technically possible at this point, the code was doing it. I have added the `canRevoke` flag back to handle this case with which this testcase another test `IncrementalCooperativeAssignorTest#testTaskAssignmentWhenWorkerBounces` which seemed to have a similar issue seem to be fixed.

Reply:
👍 Thanks. I wish we could do this without adding an extra boolean flag but preventing regressions seems worth the additional complexity, especially since there are other issues with the allocation algorithm that need to be worked out and, when fixed, will hopefully render the flag unnecessary.

Reply:
Are there any cases where the introduction of the `canRevoke` flag will defeat the purpose of this PR, by causing an imbalanced assignment to be sent out with no guaranteed follow-up rebalances (such as those caused by a scheduled rebalance delay, or workers rejoining the cluster after receiving an assignment with revocations)?

Reply:
TBH, even I didn't want to re-introduce the flag back but seemed the easiest way to get around the regression. I guess, as you said it might be easier to work through other issues on the allocation algorithm to finally have the flag redundant.

Regarding the side-effects of the re-introduction of this flag, I had imagined that adding the flag back would break some of the tests but that didn't happen which may or mayn't be a good thing. I did look at the logic again and compared with the original algorithm it seemed to me that this line:

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java#L310

is the line that prevented successive revoking rebalances. The other check here: https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java#L312 

gives us a window to set it to true and force a revocation in the next round which kind of made me believe that it should be a safe check. That said, if there are scenarios where we think we need testing, I would be happy to do that.

Reply:
Yep, I was able to produce a test case that causes an imbalanced assignment to be sent out with no revocations or delay by adding a new worker in the middle of a delayed rebalance:

```java
@Test
public void testWorkerJoiningDuringDelayedRebalance() {
    time = new MockTime();
    initAssignor();

    addNewConnector("connector3", 4);
    // First assignment with 1 worker and 3 connectors configured but not yet assigned
    performStandardRebalance();
    assertDelay(0);
    assertWorkers("worker1");
    assertConnectorAllocations(3);
    assertTaskAllocations(12);
    assertBalancedAndCompleteAllocation();

    // Second assignment with a second worker joining and all connectors running on previous worker
    // We should revoke.
    addNewEmptyWorkers("worker2");
    performStandardRebalance();
    assertWorkers("worker1", "worker2");
    assertConnectorAllocations(0, 2);
    assertTaskAllocations(0, 6);

    // Third assignment immediately after revocations, and a third worker joining.
    // This is a successive revoking rebalance. We should not perform any revocations
    // in this round
    addNewEmptyWorkers("worker3");
    performStandardRebalance();
    assertTrue(assignor.delay > 0);
    assertWorkers("worker1", "worker2", "worker3");
    assertConnectorAllocations(0, 1, 2);
    assertTaskAllocations(3, 3, 6);

    // Fourth assignment and a fourth worker joining
    // while delayed rebalance is active. We should not revoke
    time.sleep(assignor.delay / 2);
    addNewEmptyWorkers("worker4");
    performStandardRebalance();
    assertWorkers("worker1", "worker2", "worker3", "worker4");
    assertConnectorAllocations(0, 0, 1, 2);
    assertTaskAllocations(0, 3, 3, 6);

    // Fifth assignment and a fifth worker joining
    // after the delay has expired. We should revoke
    time.sleep(assignor.delay);
    addNewEmptyWorkers("worker5");
    performStandardRebalance();
    assertWorkers("worker1", "worker2", "worker3", "worker4", "worker5");
    assertNoRevocations();
    assertDelay(0);
    assertBalancedAndCompleteAllocation();
}
```

It's also worth noting that many of the test cases in this PR need to be updated with this bit at the beginning in order to mock the `Time` instance used by the assignor:
```java
time = new MockTime();
initAssignor();
```
This should be added in any test that invokes `time::sleep`.

At this point, I think we may want to split this into two separate PRs that get merged together. We can revert the `canRevoke` flag from this one, and then add a downstream PR that fixes how we calculate task-balancing revocations in tricky situations like when lost or newly-created tasks have just been assigned. That should fully address cases like the one tested [here](https://github.com/apache/kafka/blob/cda5da9b65f78b51cdfe5371f712a0d392dbdb4d/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java#L427).

@showuon WDYT? I could extract some of the changes from https://github.com/apache/kafka/pull/12019 that would fix the load-balancing revocation logic and rebase them onto this change if it helps.

Reply:
Wow, nice catch!

> At this point, I think we may want to split this into two separate PRs that get merged together. We can revert the canRevoke flag from this one, and then add a downstream PR that fixes how we calculate task-balancing revocations in tricky situations like when lost or newly-created tasks have just been assigned. That should fully address cases like the one tested [here](https://github.com/apache/kafka/blob/cda5da9b65f78b51cdfe5371f712a0d392dbdb4d/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java#L427).

Agree! Let's revert the canRevoke flag and then deal with tricky cases in a follow-up PR. Thank you.

Reply:
Thanks Chris/Luke. Sure let me revoke the `canRevoke` flag but that would mean `testTaskAssignmentWhenWorkerBounces` would fail on this PR. Also, Chris, I looked at your PR and it's closed but not merged. I see some radical changes specially =>
```
Perform load-balancing revocations any time the cluster appears imbalanced and there are still connectors and tasks that can be revoked from workers, instead of only when the number of workers in the cluster changes
```
which sounds interesting (and remember discussing this on the ticket for this) ? Also, let me know if there's anything I can also contribute to for the other PR (Looking at your comment it appeared to me that you would be rebasing the commits from that closed PR and hence the question).

Reply:
hey Chris.. Sorry to bug you again on this one. Since we anyways won't be merging this PR w/o the second one, would it make sense to incorporate the changes to this one? I believe the changes are already there in the other PR of yours?

Reply:
> Isn't this a regression? We shouldn't be revoking tasks in this round since, without those revocations, we'd have a balanced assignment.

Hello.. I have been playing around with the revokedInPrevious and the test cases. I looked at this case which is a regression which I agree. However, based on the changes made in this PR, what I think is that post the revoking rebalance, there would be a follow up rebalance which would eventually lead to a balanced load. If you look at `IncrementalCooperativeAssignorTest#testTaskAssignmentWhenWorkerBounces` this behaviour  is exhibited. On similar lines, I made some edits to `WorkerCoordinatorIncrementalTest#testTaskAssignmentWhenWorkerBounces` method by causing a follow up rebalance after the revoking rebalance after worker 3 comes back. The assignments with which I am calling the final `onLeaderElected` may not be accurate (as I couldn't get the full assignments using the deserialzeAssignment) but I just chose something which is representative.

I know this is a deviation from what the KIP proposes i.e a bounced member gets its assignments back without any revocations but if what I am stating above sounds ok, do you think it's really bad- coz it seems to be getting to a  balanced load after a a follow up rebalance. WDYT?

Reply:
Hey @vamossagar12--I'd rather err on the side of not increasing the number of rebalances required for simple cases like the one covered in `WorkerCoordinatorIncrementalTest::testTaskAssignmentWhenWorkerBounces`.

I plan to file the downstream PR sometime next week. For ease of review I'd prefer to keep it separate from this one, which has grown fairly large and complex on its own.

Reply:
Thanks @C0urante . I have reverted the test with the follow up rebalance I was basing my thoughts on the fact that KIP-415 allows having more rebalances so adding another one should be ok. Having said that, I am fine with whatever you stated as well.

Regarding a separate PR, yes that makes sense. Thanks for your help on this!

Reply:
Hi @vamossagar12--sorry for the delay. I've filed the follow-up PR here: https://github.com/vamossagar12/kafka/pull/1

I've marked it as a draft since it should not be merged into your KAFKA-12495 branch until it's approved to be merged into trunk here, but it's ready for review now. Can you take a look and see what you think? If it looks alright, we can ask Luke to take a look.

(cc @showuon)

Reply:
Thanks @C0urante ! I have added some comments for some understanding and around naming.